tensorflow/java/src/test/java/org/tensorflow/TensorMatcher.java

```diff
@@ -0,0 +1,86 @@
+/* Copyright 2016 The TensorFlow Authors. All Rights Reserved.
```

Comment:
The benefit comes from readability in the tests, and improved failure messages, and in the case of Scalar Tensors a reduction in the number of asserts needed per test. Generally, I advocate for replacing all assert* with assertThat and hamcrest matchers because of readability of code and failure messages. There are also type safety and portability (between junit and testng) benefits as well, but thats not usually why I'm advocating for them.

**Readability**
`assertThat(tensor, is(scalar());` 
versus
`assertEquals(0, tensor.numDimensions());
assertEquals(0, tensor.shape().length);`

The first when read left to right is literally _'assert that tensor is scalar'_. Using the older junit assertEquals we need two tightly coupled asserts statements to prove a tensor is scalar, which read like _'assert equals 0 the tensor.numDimensions and assert equals 0 the tensor.shape length'_

**Failure messages**
If we have a tensor with a dimension of 1 and a shape length of 1, the assertThat failure message will read:
"Expected: Tensor with shape length [0] and number of dimensions [0]
but: was Tensor with shape length [1] and number of dimensions [1]"

The assertEquals failure messages will read:
Expected: 0 was 1.

**Type safety**
assertThat(tensor, is(scalar()); will not compile if tensor is not actually a tensor. This is less of a problem for assertEquals(0, tensor.numDimension()) and assertEquals(0, tensor.shape().length) because we are calling specific function numDimension() and shape() which are non-standard java getters names anyway (so chance of accidentally having a class with the same method signatures is probably low). Still, something to think about.

Ultimately I was on the fence about adding the tensor matcher class, primarily because of the additional maintenance burden all code come with. What pushed me over to the side of adding one is that the tensor seems to be the building block of the program, so it is likely to be used all over the place to test things.